urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/Crypt.java

```diff
@@ -0,0 +1,436 @@
+                preS[j] = (char)(right[E[j] - 1] ^ KS[i][j]);
+
+            /******************************************************************
+             * The pre-select bits are now considered in 8 groups of 6 bits ea.
+             * The 8 selection functions map these 6-bit quantities into 4-bit
+             * quantities and the results are permuted to make an f(R, K).
+             * The indexing into the selection functions is peculiar;
+             * it could be simplified by rewriting the tables.
+             ******************************************************************/
+            for(j = 0; j < 8; j++) {
+                temp = 6 * j;
+                k = S[j][(preS[temp + 0] << 5) +
+                         (preS[temp + 1] << 3) +
+                         (preS[temp + 2] << 2) +
+                         (preS[temp + 3] << 1) +
+                         (preS[temp + 4] << 0) +
+                         (preS[temp + 5] << 4)];
+
+                temp = 4 * j;
+
+                f[temp + 0] = (char)((k >> 3) & 01);
+                f[temp + 1] = (char)((k >> 2) & 01);
+                f[temp + 2] = (char)((k >> 1) & 01);
+                f[temp + 3] = (char)((k >> 0) & 01);
+            }
+
+            /******************************************************************
+             * The new right is left ^ f(R, K).
+             * The f here has to be permuted first, though.
+             ******************************************************************/
+            for(j = 0; j < 32; j++)
+                right[j] = (char)(left[j] ^ f[P[j] - 1]);
+
+            /* Finally, the new left (the original right) is copied back. */
+            for(j = 0; j < 32; j++)
+                left[j] = old[j];
+        }
+
+        /* The output left and right are reversed. */
+        for(j = 0; j < 32; j++) {
+            temp = left[j];
+            left[j] = right[j];
+            right[j] = (char)temp;
+        }
+
+        /* The final output gets the inverse permutation of the very original. */
+        for(j = 0; j < 64; j++) {
+            i = FP[j];
+            if (i < 33)
+                block[j] = left[FP[j] - 1];
+            else
+                block[j] = right[FP[j] - 33];
+        }
+    }
+
+    /**************************************************************************
+     * Function:    crypt
+     *
+     * Description: Clone of Unix crypt(3) function.
+     *
+     * Inputs:      char *pw
+     *              pointer to 8 character encryption key (user password)
+     *              char *salt
+     *              pointer to 2 character salt used to modify the DES results.
+     *
+     * Returns:     Pointer to static array containing the salt concatenated
+     *              on to the encrypted results.  Same as stored in passwd file.
+     **************************************************************************/
+    private char[] crypt(char[] pw, char[] salt) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 362-367, lines 375-396, lines 403-420 into separate methods.